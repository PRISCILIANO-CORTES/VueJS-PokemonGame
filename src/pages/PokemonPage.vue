<template>
    <h1 v-if="!poke">Loading...</h1>

    <div v-else>
        <h1> Who is this Pok√©mon? </h1>
    
        <PokemonPicture :pokemonId="poke.id" :showPokemon="showPokemon" />
        <PokemonOptions :pokemon="pokemonArr" @selection="checkAnswer" />
    </div>

    <template v-if="showAnswer">
        <h2 class="fade-in"> {{ message }} </h2>
    
        <button @click="newGame">
            New game
        </button>
    </template>
</template>

<script>
    import PokemonOptions from '@/components/PokemonOptions'
    import PokemonPicture from '@/components/PokemonPicture'
    import getPokemonOptions from '@/helpers/getPokemonOptions'

    export default {
        components: { 
            PokemonOptions, 
            PokemonPicture 
        },
        data() {
            return {
                pokemonArr: [],
                poke: null,
                showPokemon: false,
                showAnswer: false,
                message: ''
            }
        },
        methods: {
            async mixPokemonArray() {
                this.pokemonArr = await getPokemonOptions()

                const rndInt = Math.floor( Math.random() * 4)

                this.poke = this.pokemonArr[ rndInt ]
            },

            checkAnswer( selectedId ) {
                this.showPokemon = true
                this.showAnswer = true

                if ( selectedId === this.poke.id ) {
                    this.message = `Correct ${ this.poke.name }`
                } else {
                    this.message = `Ups, pokemon is ${ this.poke.name }`
                }
            },
            newGame() {
                this.pokemonArr = []
                this.showPokemon = false
                this.showAnswer = false
                this.message = ''

                this.poke = null
                this.mixPokemonArray()
            }
        },
        mounted() {
            this.mixPokemonArray()
        }
    }
</script>