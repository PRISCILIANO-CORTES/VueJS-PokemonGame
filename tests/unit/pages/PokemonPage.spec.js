import { shallowMount, mount } from "@vue/test-utils"
import PokemonPage from "@/pages/PokemonPage";
import { pokemonMock } from "../mocks/pokemon.mock";

describe('PokemonPage component', () => {

    let wrapper;
    let mixPokemonArraySpy

    beforeEach( () => {
        mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' )

        wrapper = shallowMount( PokemonPage )
    })

    test('You must match the snapshot', () => {
        
        expect( wrapper.html() ).toMatchSnapshot()
    })

    test('You must call mixPokemonArray when mounting', () => {

        expect( mixPokemonArraySpy ).toHaveBeenCalled()
    })

    test('It must match the snapshot when the pokemon load', () => {
        const wrapper = mount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonMock,
                    poke: pokemonMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect( wrapper.html() ).toMatchSnapshot()
    })

    test('It should display the pokemonPicture and PokemonOptions components', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonMock,
                    poke: pokemonMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')

        expect( picture.exists() ).toBeTruthy()
        expect( options.exists() ).toBeTruthy()

        expect( picture.attributes('pokemonid') ).toBe('1')

        expect( options.attributes('pokemon') ).toBeTruthy()
    })

    test('Test in checkAnswer', async() => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonMock,
                    poke: pokemonMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(1)

        expect( wrapper.find('h2').exists() ).toBeTruthy()

        expect( wrapper.vm.showPokemon ).toBe(true)

        expect( wrapper.find('h2').text() ).toBe(`Correct ${ pokemonMock[0].name }`)

        await wrapper.vm.checkAnswer(10)

        expect( wrapper.vm.message ).toBe(`Ups, pokemon is ${ pokemonMock[0].name }`)
    })
})