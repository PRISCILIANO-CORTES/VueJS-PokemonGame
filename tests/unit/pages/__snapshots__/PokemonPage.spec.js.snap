// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component It must match the snapshot when the pokemon load 1`] = `
<div>
  <h1> Who is this Pok√©mon? </h1>
  <div class="pokemon-container"><img alt="pokemon" class="hidden-pokemon" src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg"></div>
  <div class="options-container">
    <ul>
      <li>bulbasaur</li>
      <li>ivysaur</li>
      <li>venusaur</li>
      <li>charmander</li>
    </ul>
  </div>
</div>
<!--v-if-->
`;

exports[`PokemonPage component You must match the snapshot 1`] = `
<h1>Loading...</h1>
<!--v-if-->
`;
